<a href="https://colab.research.google.com/github/miel123/DSCapstone/blob/Coursera-Data-Science-Capstone/Capstone_Project_The_Battle_of_Neighborhoods_(Week_1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1 align=center><font size = 5>The Battle of Neighborhoods</font></h1>


---
---- 



###Table of content

 [1. Introduction: Business Problem](#introduction)

 [2. Data](#Data)

 [3. Methodology](#Methodology)

 [4. Analysis](#Analysis)

 [5. Conclusion](#Conclusion)



## 1. Business Problem <a name="introduction"></a>
Covid-19 is conquering this world with uncertainty. However, geographical components of the world is not affected. While food industry is hurt by the pandemic, hospitals are needed to strive themselves to cover the increasing number of victims. Thus, the current number of hospitals is not enough, hence, they have to face more and more pressure by Covid-19. Under limited budget, gorvernment have to build more hospitals to secure people. </n>

The problem here is to find where to build new hospital(s) to efficiently handle every patient in Toronto, USA.

The problem is, where is the most suitable to build a new one. In this lab, let's leverage data analytics to solve it. 

**Solution:** 
- Eliminate locations with large number of hospitals 
- Locations must be near highly populated areas

## 2. Data <a name="Data"></a>
Based on the defined problem, factors affect the decision would be:
1. Number of existing hospitals and there location in the neighborhood
3. Ratio of hospitals/population in each neighborhood (if possible)


Thus, steps and data needed is:
- Round neiborhoods with low number of hospitals in Toronto
- Create heatmap of population in Tokyo in rounded area (if possible)
- Use K-mean cluster to pick out areas with large number of population and low number of hospitals. 
- The clustered area are candidates locations for building a new hospital.




## Methodology <a name="Methodology"></a>

Get location coordinates

 Because just extracing address in Tokyo geogegraphically may not cover the population of hospitals, we can approach by rounding from the city center 

In [1]:
#@title Set up 
!pip install pyproj
import pyproj #a coordinates library
import shapely.geometry
import folium
import pandas as pd
import numpy as np
!pip install bs4
import requests
from bs4 import BeautifulSoup
import matplotlib.pyplot as plt


     |████████████████████████████████| 6.6MB 13.3MB/s 


In [2]:
#@title Wep scrapping
link = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html_data = requests.get(link).text
soup = BeautifulSoup(html_data,'html5lib')
tables = soup.find_all('table')         
tables[0]
table_content = []
for row in tables[0].find_all('td'):
  cell = {}
  if row.span.text == 'Not assigned':
    pass
  else:
      cell['PostalCode'] = row.p.text[:3]
      cell['Borough'] = (row.span.text).split('(')[0]
      cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' / ',',')).replace(')','')).strip(' ')
      table_content.append(cell)

pd.DataFrame(table_content)


PostalCode  ...                                       Neighborhood
0          M3A  ...                                          Parkwoods
1          M4A  ...                                   Victoria Village
2          M5A  ...                           Regent Park,Harbourfront
3          M6A  ...                    Lawrence Manor,Lawrence Heights
4          M7A  ...                      Ontario Provincial Government
..         ...  ...                                                ...
98         M8X  ...        The Kingsway,Montgomery Road,Old Mill North
99         M4Y  ...                               Church and Wellesley
100        M7Y  ...                                     Enclave of M4L
101        M8Y  ...  Old Mill South,King's Mill Park,Sunnylea,Humbe...
102        M8Z  ...  Mimico NW,The Queensway West,South of Bloor,Ki...

[103 rows x 3 columns]

In [3]:
df = pd.DataFrame(table_content)


We now have the neiborhoods name without location, let's add them from the available dataset

In [4]:
#@title Fetching Coordinates
from google.colab import files
lonlat = files.upload()
import io
coord = pd.read_csv(io.BytesIO(lonlat['Geospatial_Coordinates.csv']))
df_coord = df.merge(coord, 
              left_on = 'PostalCode',
              right_on = 'Postal Code',
              how = 'left')
df_coord

Saving Geospatial_Coordinates.csv to Geospatial_Coordinates.csv


PostalCode  ...  Longitude
0          M3A  ... -79.329656
1          M4A  ... -79.315572
2          M5A  ... -79.360636
3          M6A  ... -79.464763
4          M7A  ... -79.389494
..         ...  ...        ...
98         M8X  ... -79.506944
99         M4Y  ... -79.383160
100        M7Y  ... -79.321558
101        M8Y  ... -79.498509
102        M8Z  ... -79.520999

[103 rows x 6 columns]

We now have the coordinates. Let's locate them on the map

In [5]:
address = 'Toronto,Ontario,Canada'
#!pip install Nominatim
from geopy.geocoders import Nominatim
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))
toronto_map = folium.Map(location = [latitude,longitude], zoom_start=14)

for lat,lon,name in zip(df_coord['Latitude'],df_coord['Longitude'],df_coord['Borough']): 
 folium.CircleMarker(location=[lat,lon],popup=name,tooltip='Click here to see popup', fill = True, fill_corlor = 'blue').add_to(toronto_map)
toronto_map

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


We have located our neighborhood. Now let's use Foursquare to identify hospitals

In [6]:
CLIENT_ID = 'H2PLQIRRQS1EXEXKRSXFCU33A2UIFAP3AS5SXP4KCSDIA4N0' # your Foursquare ID
CLIENT_SECRET = 'SHXBOKPUG0Q24HDXLGRKU50C0GDVLMTPMHRY12FY4GRRANFU' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)
LIMIT = 100
radius = 500
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'], 
            v['venue']['categories'][0]['name'],
            v['venue']['categories'][0]['id']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category',
                  'Category ID']
    
    return(nearby_venues)
df_venues = getNearbyVenues(df_coord['Neighborhood'],df_coord['Latitude'],df_coord['Longitude'])


Your credentails:
CLIENT_ID: H2PLQIRRQS1EXEXKRSXFCU33A2UIFAP3AS5SXP4KCSDIA4N0
CLIENT_SECRET:SHXBOKPUG0Q24HDXLGRKU50C0GDVLMTPMHRY12FY4GRRANFU
Parkwoods
Victoria Village
Regent Park,Harbourfront
Lawrence Manor,Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern,Rouge
Don MillsNorth
Parkview Hill,Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park,Princess Gardens,Martin Grove,Islington,Cloverdale
Rouge Hill,Port Union,Highland Creek
Don MillsSouth
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate,Bloordale Gardens,Old Burnhamthorpe,Markland Wood
Guildwood,Morningside,West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor,Wilson Heights,Downsview North
Thorncliffe Park
Richmond,Adelaide,King
Dufferin,Dovercourt Village
Scarborough Village
Fairview,Henry Farm,Oriole
Northwood Park,York University
The Danforth  East
Harbourfront East,Union Station,Toronto Island

## Analysis <a name="Analysis"></a>

In [7]:
df_venues['Venue'].nunique()


1251

There are 1251 venues in 0.5km near the Toronto center

We get the categories, let's filter out hospitals

In [8]:
df_hospitals = df_venues[df_venues['Category ID']=='4bf58dd8d48988d196941735']
df_hospitals

Neighborhood  ...               Category ID
1585  Kensington Market,Chinatown,Grange Park  ...  4bf58dd8d48988d196941735

[1 rows x 8 columns]

Seems like there is only one hospital in Toronto neighborhood, let's locate them on the map

In [9]:
#for lat,lon in zip(lo,df_hospitals['Venue Longitude']):
folium.Marker(location = [43.653434,-79.406074],
                      #radius = 10,
                      #color = 'red',
                      #fill = True,
                      #fill_color = '3186cc',
                      ).add_to(toronto_map)
toronto_map

Our main decision making is now moved to the population of the neighborhoods. Let's fetch population data from another source

In [10]:
link_pop ='https://en.wikipedia.org/wiki/Demographics_of_Toronto_neighbourhoods'
html_data_pop = requests.get(link_pop).text
soup = BeautifulSoup(html_data_pop,'html5lib')
tables = soup.find_all('table')         
#for i in range(0,len(tables)):
#  print("----------------- table",i,'-----------------')
#tables[1].td.td

df_neighbor = pd.DataFrame()
for i in range(1,len(tables)-4):
  data= pd.read_html(str(tables[i]),flavor = 'bs4')
  df_neighbor = df_neighbor.append(data[0])
df_neighbor

Name  ... Map
0                           Crescent Town  ... NaN
1    Governor's Bridge/Bennington Heights  ... NaN
2                                 Leaside  ... NaN
3                       O'Connor–Parkview  ... NaN
4                           Old East York  ... NaN
..                                    ...  ...  ..
151                                Woburn  ... NaN
152                              Wychwood  ... NaN
153                            York Mills  ... NaN
154               York University Heights  ... NaN
155                             Yorkville  ... NaN

[177 rows x 13 columns]

Let's map the population with postal code for each neighborhood combinations

In [11]:
df_neighbor_split = df_neighbor[['Name','Population']]
df_neighbor_split = df_neighbor_split.assign(var1 = df_neighbor_split['Name'].str.split('/')).explode('var1')

df_coord_pop = (df_coord.assign(var1=df_coord['Neighborhood'].str.split(',')).explode('var1').merge(
    df_neighbor_split,
    left_on = 'var1',
    right_on = 'var1',
    how = 'left')).dropna()
df_coord_pop

PostalCode           Borough  ...                       Name Population
0          M3A        North York  ...                  Parkwoods      26533
1          M4A        North York  ...           Victoria Village      17047
2          M5A  Downtown Toronto  ...  Regent Park/Trefann Court      10387
4          M6A        North York  ...             Lawrence Manor      13750
5          M6A        North York  ...           Lawrence Heights       3769
..         ...               ...  ...                        ...        ...
194        M4X  Downtown Toronto  ...             St. James Town      14666
195        M4X  Downtown Toronto  ...                Cabbagetown      11120
198        M8X         Etobicoke  ...               The Kingsway       8780
201        M4Y  Downtown Toronto  ...       Church and Wellesley      13397
205        M8Y         Etobicoke  ...                   Sunnylea      17602

[79 rows x 9 columns]

Now use K-means to cluster based on population

In [13]:
df0 = df_coord_pop.groupby(['Postal Code','Latitude','Longitude','Neighborhood'])['Population'].sum().reset_index()
df0

Postal Code  ...  Population
0          M1B  ...     67048.0
1          M1C  ...     36470.0
2          M1E  ...     49924.0
3          M1G  ...     48507.0
4          M1J  ...     12796.0
5          M1K  ...     13025.0
6          M1L  ...     24472.0
7          M1M  ...     23917.0
8          M1N  ...     12266.0
9          M1P  ...     14189.0
10         M1R  ...     26644.0
11         M1S  ...     44577.0
12         M1V  ...     26272.0
13         M2J  ...      2790.0
14         M2K  ...     12280.0
15         M2L  ...     17564.0
16         M2M  ...     79190.0
17         M3A  ...     26533.0
18         M3H  ...     28677.0
19         M4A  ...     17047.0
20         M4E  ...     20416.0
21         M4G  ...     13876.0
22         M4H  ...     17949.0
23         M4K  ...     31007.0
24         M4N  ...      6653.0
25         M4S  ...     23727.0
26         M4T  ...      4474.0
27         M4V  ...     21383.0
28         M4W  ...      7672.0
29         M4X  ...     25786.0
30         M4Y  ...     13397.0
31         M5A  ...     10387.0
32         M5B  ...      8240.0
33         M5C  ...     14666.0
34         M5J  ...       627.0
35         M5M  ...     13749.0
36         M5R  ...     21647.0
37         M5T  ...     12747.0
38         M6A  ...     17519.0
39         M6J  ...      5013.0
40         M6K  ...      9039.0
41         M6M  ...     21284.0
42         M6N  ...      4382.0
43         M6R  ...     44363.0
44         M6S  ...     15515.0
45         M8V  ...     21230.0
46         M8W  ...     21281.0
47         M8X  ...      8780.0
48         M8Y  ...     17602.0
49         M9B  ...     13683.0
50         M9C  ...     10240.0
51         M9L  ...     12766.0
52         M9M  ...      4327.0
53         M9N  ...     16476.0
54         M9P  ...      5857.0
55         M9R  ...     16254.0
56         M9V  ...     16790.0
57         M9W  ...     15825.0

[58 rows x 5 columns]

In [19]:
!pip install sklearn
from sklearn.cluster  import KMeans
kmeans = KMeans(n_clusters= 5,random_state = 4).fit(df0[['Population']])
df0['Cluster'] = kmeans.labels_

In [23]:
import matplotlib.cm as cm
import matplotlib.colors as colors
# set 5 clusters
# set color scheme for the clusters
x = np.arange(5)
ys = [i + x + (i*x)**2 for i in range(5)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(df0['Latitude'], df0['Longitude'], df0['Neighborhood'], df0['Cluster']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=10,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(toronto_map)
       
toronto_map

## Conclusion <a name = "Conclusion"> </a> 

Based on the clustering, we can place new hospital in area with red dot (highly populated areas)

Propose with the investors and stake holders to further decide the locations for the new hospital. Other elements should be taken into consideration are budget, time, and number of covid-19 patients currently have. 